### PR TITLE
Add `start_day` parameter to `Time#at_beginning_of_week`

### PR DESCRIPTION
**Pull request by DanielGilchrist**
Closes https://github.com/crystal-lang/crystal/issues/13447

Currently when using `Time#at_beginning_of_week` it assumes the start of the week to be Monday. This works great for many use cases but sometimes we may want to assume the beginning of the week as another day (e.g. Sunday is also commonly considered to be "the beginning of the week")

~I've opted to allow this in 2 forms: With a `Time::DayOfWeek` which is type safe and for convenience, a `String` or `Symbol` that raises an `ArgumentError` if an invalid value is passed~ 
**EDIT:** @Blacksmoke16 pointed out [here](https://github.com/crystal-lang/crystal/pull/13446#issuecomment-1538794685) that we can leverage Symbol autocasting for enum members and avoid the overload altogether

```cr
Time.local.at_beginning_of_week # => 2023-05-08 00:00:00 +00:00
Time.local.at_beginning_of_week(Time::DayOfWeek::Sunday) # => 2023-05-07 00:00:00 +00:00
Time.local.at_beginning_of_week(:sunday) # => 2023-05-07 00:00:00 +00:00
```

This kind of functionality is provided by Rails which is where I got the inspiration from https://api.rubyonrails.org/classes/DateAndTime/Calculations.html#method-i-beginning_of_week

If we decide this is worth having in the standard library I'd also be more than happy to implement it for `Time#at_end_of_week` in this PR or a separate one as well :)

### Current
image
image

### After this PR
image
image

Thanks!

### CODE REVIEW

**Comment by Blacksmoke16** [comment]
Due to https://crystal-lang.org/reference/1.8/syntax_and_semantics/autocasting.html#symbol-autocasting, I' probably be in favor of dropping support for `String` and just leverage the Eum + Symbol autocasting. Then wouldn't even need the extra overload.

**Comment by DanielGilchrist** [comment]
> Due to [crystal-lang.org/reference/1.8/syntax_and_semantics/autocasting.html#symbol-autocasting](https://crystal-lang.org/reference/1.8/syntax_and_semantics/autocasting.html#symbol-autocasting), I'd probably be in favor of dropping support for `String` and just leverage the Enum plus Symbol autocasting. Then wouldn't even need the extra overload.

That's awesome! Thanks for pointing that out. I'll put a commit together now

File: src/time.cr
Comment (Sija) [inline src/time.cr:1374]: You can use auto-casting already at the definition.

```suggestion
  def at_beginning_of_week(start_day : Time::DayOfWeek = :monday) : Time
```

Comment (DanielGilchrist) [reply src/time.cr]: Thanks!

**Comment by straight-shoota** [comment]
Thanks for your contribution, @DanielGilchrist. 🙇
It's very appreciated that you take the initiative. Experience tells that discussions about feature enhancements are better discussed independently of a concrete implementation. Because of this, [the contribution guide](https://github.com/crystal-lang/crystal/blob/master/CONTRIBUTING.md#using-the-issue-tracker) asks to open an issue before sending a pull request.

Now we already have PR and that's great. I'd still ask you to open an issue to propose this change (you can basically take the PR description for that), so we can discuss the feature there.
So long this PR can stay open (ideally in *draft* mode) and we'll get back here when the feature addition is agreed upon.

**Comment by DanielGilchrist** [comment]
> Thanks for your contribution, @DanielGilchrist. 🙇 It's very appreciated that you take the initiative. Experience tells that feature enhancements are better discussed independently of a concrete implementation. Because of this, [the contribution guide](https://github.com/crystal-lang/crystal/blob/master/CONTRIBUTING.md?rgh-link-date=2023-05-08T18%3A50%3A27Z#using-the-issue-tracker) asks to open an issue before sending a pull request.
> 
> Now we already have PR and that's great. I'd still ask you to open an issue to propose this change (you can basically take the PR description for that), so we can discuss the feature there. So long this PR can stay open (ideally in _draft_ mode) and we'll get back here when the feature addition is agreed upon.

Thank you for your response and apologies for not reading the contribution guide thoroughly enough 😅 I'll make sure to give it a good read for future contributions

I'll put this PR into draft mode and make an issue now :)

Thanks!

File: src/time.cr
Comment (Sija) [inline src/time.cr:1373]: IMO there's no need to separately showcase the autocasting feature, I'd suggest to simply show it's possible.

```suggestion
  # ```
  # now = Time.local                  # => 2023-05-08 16:48:13 +00:00
  # now.at_beginning_of_week(:monday) # => 2023-05-08 00:00:00 +00:00
  # now.at_beginning_of_week(:sunday) # => 2023-05-07 00:00:00 +00:00
  # ```
  #
  # TODO: Ensure correctness in local time-line.
```

**Comment by straight-shoota** [comment]
https://github.com/crystal-lang/crystal/pull/13446 has had a couple of upvotes (including from two Core Team members), so I think we're good to move on. Sorry for the inconvenience and thanks for following along🙇
This was luckily an uneventful proposal process =)

Simplifying the documentation as suggested in https://github.com/crystal-lang/crystal/pull/13446#discussion_r1187927855 would be good. Then this should be ready to roll.

File: spec/std/time/time_spec.cr
Comment (HertzDevil) [inline spec/std/time/time_spec.cr:1]: There are no specs to show that the hour/minute/second components become 0 after a call to `#at_beginning_of_week`

**Review by HertzDevil — CHANGES_REQUESTED** [review]

File: src/time.cr
Comment (DanielGilchrist) [reply src/time.cr]: Thank you! I've addressed this here: [`a08f0d2` (#13446)](https://github.com/crystal-lang/crystal/pull/13446/commits/a08f0d22e4832fc5b8dfb395406eac96269007d4)

I also opted to add an example showing the default behaviour without an argument :~)

**Comment by DanielGilchrist** [comment]
> #13446 has had a couple of upvotes (including from two Core Team members), so I think we're good to move on. Sorry for the inconvenience and thanks for following along🙇 This was luckily an uneventful proposal process =)

No worries and thank you for your time!

> Simplifying the documentation as suggested in [#13446 (comment)](https://github.com/crystal-lang/crystal/pull/13446#discussion_r1187927855) would be good. Then this should be ready to roll.

I've addressed the remaining comments and relevant specs are passing locally let me know if there are any other changes required before merge :~)

**Review by HertzDevil — APPROVED** [review]

File: src/time.cr
Comment (straight-shoota) [inline src/time.cr:1368]:
```suggestion
  # now = Time.utc(2023, 5, 16, 17, 53, 22)
  # now.at_beginning_of_week             # => 2023-05-15 00:00:00 UTC
  # now.at_beginning_of_week(:sunday)    # => 2023-05-14 00:00:00 UTC
  # now.at_beginning_of_week(:wednesday) # => 2023-05-10 00:00:00 UTC
```

**Review by straight-shoota — COMMENTED** [review]
I have a small proposal to improve reproduceability of the example by using a fixed timestamp instead of `Time.local`.

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**
